File: test-app/runtime/src/main/cpp/WorkerEvents.cpp
#include "WorkerEvents.h"

#include "ArgConverter.h"
#include "BuiltinLoader.h"
#include "NativeScriptException.h"
#include "Runtime.h"
#include "RuntimeState.h"

using namespace v8;

namespace tns {

namespace {

/*
 * The worker-events builtin's delivery callouts for this isolate. Both message
 * directions share emitMessage; only the receiver differs. emitError is
 * parent-side only.
 */
struct WorkerEventsState {
    Global<v8::Function> emitMessage;
    Global<v8::Function> emitError;
};

Local<v8::Function> CalloutOf(Local<Object> exports, Isolate* isolate, const char* name) {
    Local<Value> callout;
    if (!exports->Get(isolate->GetCurrentContext(),
                      ArgConverter::ConvertToV8String(isolate, name))
                 .ToLocal(&callout) ||
        !callout->IsFunction()) {
        throw NativeScriptException(std::string("WorkerEvents::Init: the worker-events "
                                                "bootstrap did not return ") +
                                    name);
    }
    return callout.As<v8::Function>();
}

}  // namespace

void WorkerEvents::Init(Local<Context> context) {
    Isolate* isolate = v8::Isolate::GetCurrent();

    Local<Object> exports;
    if (!BuiltinLoader::GetExports(context, BuiltinId::kWorkerEvents, nullptr)
                 .ToLocal(&exports)) {
        throw NativeScriptException("WorkerEvents::Init: the worker-events bootstrap failed");
    }

    Local<v8::Function> emitMessage = CalloutOf(exports, isolate, "emitMessage");
    Local<v8::Function> emitError = CalloutOf(exports, isolate, "emitError");

    auto* state = RuntimeState::For<WorkerEventsState>(isolate);
    if (state == nullptr) {
        throw NativeScriptException("WorkerEvents::Init: no runtime state for isolate");
    }
    state->emitMessage.Reset(isolate, emitMessage);
    state->emitError.Reset(isolate, emitError);
}

void WorkerEvents::EmitMessage(Isolate* isolate, Local<Object> receiver,
                               const std::shared_ptr<worker::Message>& message) {
    auto* state = RuntimeState::For<WorkerEventsState>(isolate);
    if (state == nullptr || state->emitMessage.IsEmpty()) {
        return;
    }
    Runtime* runtime = Runtime::TryGetRuntime(isolate);
    if (runtime == nullptr) {
        return;
    }
    Local<Context> context = runtime->GetContext();

    Local<Value> data;
    Local<Value> ports;
    const char* type = "message";
    {
        TryCatch tc(isolate);
        if (!message->Deserialize(isolate, context, &ports).ToLocal(&data)) {
            if (tc.HasTerminated()) {
                return;
            }
            // HTML: a message that cannot be read still reaches its target, as
            // a `messageerror` event carrying nothing.
            tc.Reset();
            data = v8::Undefined(isolate);
            ports = Local<Value>();
            type = "messageerror";
        }
    }

    Local<Value> args[3]{data,
                         ports.IsEmpty() ? v8::Undefined(isolate).As<Value>() : ports,
                         ArgConverter::ConvertToV8String(isolate, type)};
    Local<Value> result;
    // A throw here is left pending on purpose: on the worker side the drain's
    // TryCatch turns it into the scope's error event, and on the parent side
    // the internal-lane entry runs with no TryCatch of its own, so V8's
    // uncaught-message listener reports it.
    (void)state->emitMessage.Get(isolate)->Call(context, receiver, 3, args).ToLocal(&result);
}

bool WorkerEvents::EmitError(Isolate* isolate, Local<Object> receiver,
                             const std::string& message, const std::string& source,
                             const std::string& stackTrace, int lineNumber) {
    auto* state = RuntimeState::For<WorkerEventsState>(isolate);
    if (state == nullptr || state->emitError.IsEmpty()) {
        return false;
    }
    Runtime* runtime = Runtime::TryGetRuntime(isolate);
    if (runtime == nullptr) {
        return false;
    }
    Local<Context> context = runtime->GetContext();

    Local<Value> args[4]{ArgConverter::ConvertToV8String(isolate, message),
                         ArgConverter::ConvertToV8String(isolate, source),
                         Number::New(isolate, lineNumber),
                         ArgConverter::ConvertToV8String(isolate, stackTrace)};
    Local<Value> result;
    if (!state->emitError.Get(isolate)->Call(context, receiver, 4, args).ToLocal(&result)) {
        return false;
    }
    return result->BooleanValue(isolate);
}

}  // namespace tns

File: test-app/runtime/src/main/cpp/js/worker-events.js
"use strict";
// Worker (HTML Standard §10.2.6) and the worker global scope (§10.2.1) as
// EventTargets: both deliver MessageEvents instead of the runtime's historical
// direct call of an `onmessage` property, and the Worker object receives the
// worker's unhandled errors as ErrorEvents. The worker global scope's own
// `onerror` stays a direct call with the error — a documented NativeScript
// contract, not the web's event.
//
// Eager, because the handler attributes have to exist before app code assigns
// one. MessageEvent itself is pulled in on the first delivery, so a worker
// nobody talks to never runs that builtin.
const { ObjectDefineProperty, ObjectSetPrototypeOf } = primordials;

const {
  Event,
  EventTarget,
  defineEventHandler,
  dispatchEventRethrowing,
  globalEventTarget,
} = require("internal/events");

const g = globalThis;

let MessageEvent;
function getMessageEvent() {
  if (MessageEvent === undefined) {
    ({ MessageEvent } = require("internal/message-event"));
  }
  return MessageEvent;
}

// ErrorEvent is installed by the error-events builtin, which
// Runtime::PrepareV8Runtime runs AFTER this one — so the constructor can
// only be taken on the first error delivery, not at init.
let ErrorEvent;
function getErrorEvent() {
  if (ErrorEvent === undefined) {
    ErrorEvent = g.ErrorEvent;
  }
  return ErrorEvent;
}

// The delivery callout, invoked by native with the receiving target as `this`:
// the Worker object on the parent isolate, the global scope's EventTarget
// inside a worker. `ports` is the array of MessagePorts the message
// transferred, or undefined when it carried none.
//
// A handler that throws propagates back into the calling native frame: that
// is what feeds the worker's onerror chain — the worker scope's handler
// first, then the parent's — which the cross-runtime worker suite asserts.
function emitMessage(data, ports, type) {
  const MessageEventCtor = getMessageEvent();
  dispatchEventRethrowing(this, new MessageEventCtor(type, { data, ports }));
}

// The parent-side error delivery callout, invoked by native with the Worker
// object as `this` once the worker scope has left the error unhandled. Only
// primitives cross the isolate boundary, so the event carries no `error`
// object; `stackTrace` is this runtime's addition to the ErrorEvent fields.
//
// Returns whether the error was handled: a truthy return from the `onerror`
// attribute cancels the event (HTML §8.1.7.3), as does preventDefault() from
// any listener.
function emitError(message, filename, lineno, stackTrace) {
  const ErrorEventCtor = getErrorEvent();
  const event = new ErrorEventCtor("error", {
    message,
    filename,
    lineno,
    cancelable: true,
  });
  event.stackTrace = stackTrace;
  dispatchEventRethrowing(this, event);
  return event.defaultPrevented;
}

// The parent-side end-of-worker callout, invoked by native with the Worker
// object as `this` once the worker's thread has finished — its own close() as
// much as a terminate(). `nsworkerended` is internal and non-standard: the web
// has no end-of-worker event, and the node:worker_threads shim is what turns
// this into an 'exit'.
function emitEnded() {
  dispatchEventRethrowing(this, new Event("nsworkerended"));
}

ObjectSetPrototypeOf(g.Worker.prototype, EventTarget.prototype);
defineEventHandler(g.Worker.prototype, "message");
defineEventHandler(g.Worker.prototype, "messageerror");
defineEventHandler(g.Worker.prototype, "error", "error", true);

// The global scope's handler attributes are defined against the EventTarget
// backing the global listener methods, which is what native dispatches on and
// what globalThis.addEventListener registers with — so a handler and an
// addEventListener registration interleave in assignment order. globalThis
// only forwards.
defineEventHandler(globalEventTarget, "message");
defineEventHandler(globalEventTarget, "messageerror");
for (const name of ["onmessage", "onmessageerror"]) {
  ObjectDefineProperty(g, name, {
    __proto__: null,
    get() {
      return globalEventTarget[name];
    },
    set(value) {
      globalEventTarget[name] = value;
    },
    enumerable: true,
    configurable: true,
  });
}

module.exports = { emitMessage, emitError, emitEnded };
